src/extension.rs

```diff
@@ -0,0 +1,881 @@
+use crate::alloc::boxed::Box;
+use crate::alloc::collections::btree_map::Entry;
+use crate::alloc::collections::BTreeMap;
+use crate::alloc::fmt::{Display, Formatter};
+use crate::encoding::{DecodeContext, WireType};
+use crate::generic::{EncodeBuffer, Merge, MergeBuffer, ProtoIntType};
+use crate::{DecodeError, Encode};
+use bytes::{Buf, BufMut};
+use core::any::Any;
+use core::fmt::Debug;
+use core::marker::PhantomData;
+use core::option::Option;
+
+type ExtendableTypeId = &'static str;
+type FieldTag = u32;
+
+/// Marks a message as having able to have extensions.
+/// Extension data must be retrieved and modified through use of static Extensions in generated code.
+pub trait Extendable: 'static {
+    /// A static type id associated with this Extendable.
+    fn extendable_type_id() -> ExtendableTypeId;
```

Comment:
```suggestion
    fn extendable_type_id(&self) -> &'a str;
```
Could this return a borrowed string to allow implementing it for something like [`DynamicMessage`](https://docs.rs/prost-reflect/0.8.1/prost_reflect/struct.DynamicMessage.html)? Most consumers can just call `T::default().extendable_type_id()`